hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationThread.java

```diff
@@ -118,9 +118,27 @@ MigrationRunnable getActiveTask() {
         return activeTask;
     }
 
+    /**
+     * Interrupts the migration thread and joins on it.
+     * <strong>Must not be called on the migration thread itself</strong> because it will result in infinite blocking.
+     */
     void stopNow() {
+        assert currentThread() != this : "stopNow must not be called on the migration thread";
         queue.clear();
         interrupt();
+        boolean currentThreadInterrupted = false;
+        while (true) {
```

Comment:
Actually it's not, they are from a strictly controlled set of `MigrationRunnable`s. However, given our overall interruptibility status this might be fragile.